### PR TITLE
fix(fmgc): handle precision faf constraints correctly

### PR DESCRIPTION
**Pull request by tracernz**
Fixes #[issue_no]

## Summary of Changes

Constraints on precision approach FAFs in MSFS navdata are coded incorrectly, therefore they are treated as, and displayed as regular constraints in the aircraft. We fix up the the constraint coding, and handle it properly so the altitudes are not applied, as per the real aircraft, and additionally use them to display the glideslope intercept altitude on the destination VERT REV page.

Also fixes another small bug in the nav database that causes a lot of issues on sentry.

## Screenshots (if necessary)

Before:
image
After:
image
image
image

## References

ARINC429
Airbus docs
Pilots (special thanks Eaglepilot6 for pointing this out)
vert-rev-arrival-runway
image

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Check that GS intercept on the destination VERT REV page is correct for a few different ILS approaches. Check that no altitude constraint is applied on the FAF for ILS approaches (they should still be applied for non-precision approaches). Check that destination VERT REV QNH field matches up with PERF APPR. Check that vert rev page works fine on all other waypoints.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js [hunk 1]
```diff
@@ -313,6 +313,7 @@ class CDUFlightPlanPage {
                 let timeColor = color;
 
                 // Altitude
+                const hasAltConstraint = wp.legAltitudeDescription > 0 && wp.legAltitudeDescription < 6;
```
Comment (BlueberryKing) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:316]: Why do we check for `wp.legAltitudeDescription < 6` here? It looks like this is the condition to display `legAltitude1` in the flight plan. In the case of `wp.legAltitudeDescription < 6`, the relevant constraint is stored in `legAltitude1` according to this: https://github.com/flybywiresim/a32nx/blob/master/src/fmgc/src/types/fstypes/FSEnums.ts#L47

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js [hunk 2]
```diff
@@ -341,14 +342,18 @@ class CDUFlightPlanPage {
...
-                        altitudeConstraint = wp.legAltitude1.toFixed(0).toString().padStart(5,"\xa0");
+                    // TODO all this constraint code is very sussy
+
+                    if (hasAltConstraint) {
+                        if (mcdu.flightPlanManager.getOriginTransitionAltitude() >= 100 && wp.legAltitude1 > mcdu.flightPlanManager.getOriginTransitionAltitude()) {
+                            altitudeConstraint = (wp.legAltitude1 / 100).toFixed(0).toString();
+                            altitudeConstraint = `FL${altitudeConstraint.padStart(3,"0")}`;
+                        } else {
+                            altitudeConstraint = wp.legAltitude1.toFixed(0).toString().padStart(5,"\xa0");
+                        }
```
Comment (BlueberryKing) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:353]: Just some formatting suggestions for the sake of consistency. I know this is how it was previously, but it's gonna bother me if I see it like this later 😄.

```suggestion
                    if (hasAltConstraint) {
                        if (mcdu.flightPlanManager.getOriginTransitionAltitude() >= 100 && wp.legAltitude1 > mcdu.flightPlanManager.getOriginTransitionAltitude()) {
                            altitudeConstraint = (wp.legAltitude1 / 100).toFixed(0).toString();
                            altitudeConstraint = `FL${altitudeConstraint.padStart(3, "0")}`;
                        } else {
                            altitudeConstraint = wp.legAltitude1.toFixed(0).toString().padStart(5, "\xa0");
                        }
```

File: src/fmgc/src/flightplanning/LegsProcedure.ts [hunk 3]
```diff
@@ -200,18 +201,26 @@ export class LegsProcedure {
...
                   mappedLeg.speedConstraint = currentLeg.speedRestriction;
                   mappedLeg.turnDirection = currentLeg.turnDirection;
                   mappedLeg.additionalData.legType = currentLeg.type;
                   mappedLeg.additionalData.overfly = currentLeg.flyOver;
+                  mappedLeg.additionalData.fixTypeFlags = currentLeg.fixTypeFlags;
 
                   mappedLeg.additionalData.distance = currentLeg.distanceMinutes ? undefined : currentLeg.distance / 1852;
                   mappedLeg.additionalData.distanceInMinutes = currentLeg.distanceMinutes ? currentLeg.distance : undefined;
                   mappedLeg.additionalData.course = currentLeg.trueDegrees ? currentLeg.course : A32NX_Util.magneticToTrue(currentLeg.course, Facilities.getMagVar(mappedLeg.infos.coordinates.lat, mappedLeg.infos.coordinates.long));
-                  mappedLeg.additionalData.overfly = currentLeg.flyOver;
```
Comment (BlueberryKing) [inline src/fmgc/src/flightplanning/LegsProcedure.ts:214]: Out of curiosity, why is this change required?

**Review by BlueberryKing — APPROVED** [review]
Cool stuff! This is important groundwork for some more VNAV functionality. I left a couple of comments if you wouldn't mind. It's just questions that came up.

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js (hunk 1 above)
Comment (tracernz) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:316]: Yes, the data is not a constraint for precision approaches (descriptors >= 6 are used for those).

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js (hunk 2 above)
Comment (tracernz) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:353]: I think we will replace this code entirely in cfms-20, but can do this now. :D

File: src/fmgc/src/flightplanning/LegsProcedure.ts (hunk 3 above)
Comment (tracernz) [reply src/fmgc/src/flightplanning/LegsProcedure.ts:214]: Duplicate... it's already done a few lines above. 😆

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #6933
Tier of Testing  : 1
Date             : 21/03/2022

Testing Process: 
Check that GS intercept on the destination VERT REV page is correct for a few different ILS approaches. - **OK**
Check that no altitude constraint is applied on the FAF for ILS approaches (they should still be applied for non-precision approaches).  - **OK**
Check that destination VERT REV QNH field matches up with PERF APPR. - **OK**
Check that vert rev page works fine on all other waypoints. - **OK**

Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
LGTM

Media:       
image
